### ISSUE TITLE
WEB TODO

### ISSUE THREAD

**Issue by sh-dv** [covert-encryption/covert#56]
This is the to-do list for the first JS prototype of Covert.
This prototype should demonstrate fast and efficient encryption of files/messages using chunked encryption (with chacha20poly1305 and ARGON2 browser SIMD) and service worker stream based file download (FetchEvent with ReadableStream based Response to download). And public key encrypt & sign with SSH, Age and Minisign keys.

- [ ] Blockstream - encrypt_file/decrypt_file, header encryption, chacha implementation
- [ ] Archive + Armor encode/decode
- [ ] elligator
- [ ] passphrase generation + wordlist + zxcvbn
- [ ] PublicKey : reading, decode, minisign,ssh,age,gh + signing
- [ ] xor, random padding, nonce generation, bech32 checksum.

This should be interesting.
If something is missing feel free to edit. Or if it looks messy/unorganized lol.

**Comment by covert-encryption** [comment]
XOR only needs a simple implementation that also happens to run quite fast (not that performance is relevant to Covert here):
```javascript
function xor(a, b) {
  const a32 = new Uint32Array(a), b32 = new Uint32Array(b)  // These are ArrayBuffer views, no copying
  for (let i = 0; i < b32.length; ++i) a32[i] ^= b32[i]
}
```

**Comment by covert-encryption** [comment]
Decryption of Minisign keys needs scrypt and blake2b and SSH needs bcrypt and aes, none of which are used anywhere else in Covert. AES is part of WebCrypto/Subtle and thus available in browsers, others need JS/Wasm implementation but the only one that needs to be fast is scrypt. Other operations take almost no time even with very inefficient implementation. SSH uses extremely weak settings with bcrypt. Also, SSH keys without password need none of the crypto, but Minisign is always encrypted even with an empty passphrase, taking nearly one second to calculate on native implementation.

For an initial implementation, support for these encrypted keys can easily be omitted.

**Comment by covert-encryption** [comment]
Armor encode/decode can use standard Base64 functions already available in browsers. Python Covert adds some additional processing that is not crucial to implement, and that can be ported quite easily to Javascript if so desired.

**Comment by sh-dv** [comment]
In another issue there was mentioned that a previous similar project was done in JS. is there any github link to that or a gist ? Also in the python code there are some cffi lines of code, is that something that should be taking into account in js ? Also the bytes() function kinda tricky in js. These are somethings i noticed when i took a look at the covert python code.

**Comment by covert-encryption** [comment]
@sh-dv chacha.py reimplements some Python code of pynacl, which itself does not properly expose the C functions it wraps. We do not compile any C code in Covert, just using the existing raw C bindings provided by pynacl (and this is really only for maximal speed, avoiding some buffer copying). Javascript ports of libsodium probably don't need any such changes.

You can simply `new ArrayBuffer(n)` to accomplish the same as `bytes(n)`. Javascript ArrayBuffer has the benefit of being read-write, while Python bytes are read-only (thus calling for use of read-write `bytearray` and some conversions in various places where `bytes` are used, and additionally some conversions between int/bytes where numbers are needed in cryptography and for blockstream nextlen fields). For `memoryview` there is a similar view JS construct `new DataView(...)` that avoids copying of data on slicing (where `buf.slice(...)` always makes a copy and that can be expensive if the slice is large).

I could have a look at writing most of the worker code in JS myself, if you are interested on handling the other parts. And if you do plan a bigger rewrite of hat.sh at some point, you are perfectly welcome to base it on the Covert format.

**Comment by sh-dv** [comment]
That sounds perfect.
It will be much easier if i took a deep look at pynacl and compare it to the C or JS ones, then i will have better overall understanding.

**Comment by covert-encryption** [comment]
Do you know if it is possible to run multiple workers that run in separate threads using more than one CPU core? I know that a single worker is always single-threaded but it would be awesome if we could do some things in parallel for extra speed.

**Comment by sh-dv** [comment]
Depends on what type of workers. Is it a web worker or a service worker?

In web workers you can run multiple workers without any issues. Multi threading can be achieved. and full utilization of cpu cores. **However**, web workers can't handle [fetch events](https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent) and responses, thus you are limited to low file sizes (RAM usage). Or just wait for the [FileSystemWritableFileStream](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemWritableFileStream#browser_compatibility) feature to be released and fully supported in all major browsers. which in my opinion behave the same way hat.sh sw does.

In service workers, i think it is the same case regarding multi threading and cpu usage. because both workers are features provided by browsers and can be used through JS. SW [life cycle](https://developers.google.com/web/fundamentals/primers/service-workers/lifecycle) have to be taken into account. 
In my experience while developing v2 beta of hat.sh, i tried to do some concurrency to make things go faster and that went horribly. Not only that, trying the same code (encryption) in 2 windows (tabs) in the same browser went bad and file mix up happened. Because it is a service worker (fetch event).

What you are trying to accomplish here could be done in a web worker without any issues (afaik). However there is the file size limit because it will be read as a whole in memory, unless you can find another way to stream download, which I spent a lot of time figuring that out and failed and chose service workers.

**Comment by covert-encryption** [comment]
That is good to hear. I believe that most work should indeed be done in web workers. Data can be passed around in `SharedArrayBuffer` so the file size should not be an obstacle, and for password hashing only small messages need to be sent. However, I realise that such implementation is much more difficult and thus would first go for single threaded (threads are the reason for most of the complexity you see in blockstream.py).

Just considering the future options here, and possibly FileSystemWritableFileStream will also be supported soon enough :)

**Comment by covert-encryption** [comment]
Updated TODO on first post. Not everything needs to be implemented at first, but I believe that feature-wise it is quite complete now. This is supposed to be implemented **roughly** from top to bottom, both on the main components and on the subtasks on each point.

I would start by getting decryption working for password-encrypted files that only contain one block (i.e. are 1024 bytes or smaller). Once that is working, it is good to build up from there by making blockstream work, and once that runs OK for both encryption and decryption, look at implementing the archive. And only finally going for public keys, given how many points there are to be solved for them. And signatures should be left pretty much last, as not even the Python side specification on those is finished.

**Comment by covert-encryption** [comment]
Password normalise/UTF-8 solved:
```javascript
// String to ArrayBuffer conversion with Unicode normalisation
const encode = str => new TextEncoder().encode(str.normalize('NFKC'))
```

**Comment by covert-encryption** [comment]
I think the best that can be done for github keys is to open the link in new tab and ask the user to copy&paste. They block everything else via CORS and frame policy.

**Comment by sh-dv** [comment]
Cool. Do you plan on opening a new repo named covert-web or a new folder in the main repo?

**Comment by covert-encryption** [comment]
We can fork one from your project and give you permissions to it, or you can make a branch on yours, whichever you prefer. A separate repo of Python version in any case, as no code is shared between the two.

**Comment by sh-dv** [comment]
I suggest a new repo "covert-web" on your acc.
First, a JS demo should be established once necessary tasks above are completed.
Then that demo(prototype) will be converted in a modern react ui.
